lib/SourceMapDevToolPlugin.js

```diff
@@ -163,233 +161,279 @@ class SourceMapDevToolPlugin {
-								 */
-								if (cachedFile !== file) {
-									const chunk = fileToChunk.get(file);
-									if (chunk !== undefined) chunk.auxiliaryFiles.add(cachedFile);
-								}
-							}
-							return;
-						}
+					let fileIndex = 0;
+
+					asyncLib.each(
+						files,
+						(file, done) => {
+							const asset = compilation.getAsset(file).source;
+							const cacheIdent = `${compilation.compilerPath}/SourceMapDevToolPlugin/${file}`;
+							const cacheETag = createHash("md4");
+							asset.updateHash(cacheETag);
 
-						reportProgress(
-							(0.5 * idx) / files.length,
-							file,
-							"generate SourceMap"
-						);
-						/** @type {SourceMapTask | undefined} */
-						const task = getTaskForFile(file, asset, options, compilation);
-
-						if (task) {
-							const modules = task.modules;
-
-							for (let idx = 0; idx < modules.length; idx++) {
-								const module = modules[idx];
-								if (!moduleToSourceNameMapping.get(module)) {
-									moduleToSourceNameMapping.set(
-										module,
-										ModuleFilenameHelpers.createFilename(
-											module,
-											{
-												moduleFilenameTemplate: moduleFilenameTemplate,
-												namespace: namespace
-											},
-											{
-												requestShortener,
-												chunkGraph
+							compilation.cache.get(
+								cacheIdent,
+								cacheETag.digest("hex").toString(),
+								(err, assets) => {
+									/**
+									 * If presented in cache, reassigns assets. Cache assets already have source maps.
+									 */
+									if (assets) {
+										for (const cachedFile in assets) {
+											if (cachedFile === file) {
+												compilation.updateAsset(cachedFile, assets[cachedFile]);
+											} else {
+												compilation.emitAsset(cachedFile, assets[cachedFile], {
+													development: true
+												});
 											}
-										)
-									);
-								}
-							}
+											/**
+											 * Add file to chunk, if not presented there
+											 */
+											if (cachedFile !== file) {
+												const chunk = fileToChunk.get(file);
+												if (chunk !== undefined)
+													chunk.auxiliaryFiles.add(cachedFile);
+											}
+										}
 
-							tasks.push(task);
-						}
-					});
+										done();
 
-					reportProgress(0.5, "resolve sources");
-					/** @type {Set<string>} */
-					const usedNamesSet = new Set(moduleToSourceNameMapping.values());
-					/** @type {Set<string>} */
-					const conflictDetectionSet = new Set();
+										return;
+									}
 
-					/**
-					 * all modules in defined order (longest identifier first)
-					 * @type {Array<string | Module>}
-					 */
-					const allModules = Array.from(moduleToSourceNameMapping.keys()).sort(
-						(a, b) => {
-							const ai = typeof a === "string" ? a : a.identifier();
-							const bi = typeof b === "string" ? b : b.identifier();
-							return ai.length - bi.length;
-						}
-					);
+									reportProgress(
```

Comment:
As this is no longer sync, progress reporting should be moved to the bottom, just before the `done` call.